File: android/app/src/main/java/org/cashu/wallet/Core/SentryService.kt
package org.cashu.wallet.Core

import android.content.Context
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryLevel
import io.sentry.android.core.SentryAndroid

internal interface SentryGateway {
    fun start(dsn: String)
    fun close()
    fun capture(error: Throwable)
    fun breadcrumb(message: String, category: String)
}

/**
 * Opt-in crash reporting mirroring Swift `SentryService`: every entry point except
 * `shutdown()` is a no-op unless the user enabled `settings.sentryEnabled` (default off).
 */
class SentryService internal constructor(
    private val gateway: SentryGateway,
    private val isEnabled: () -> Boolean,
) {
    constructor(context: Context, settingsStore: SettingsStore) : this(
        gateway = AndroidSentryGateway(context.applicationContext),
        isEnabled = { settingsStore.sentryEnabled },
    )

    companion object {
        // Shared with iOS (SentryService.swift) — one Sentry project for both platforms.
        private const val DSN =
            "https://aff293071a9e53305e76990761d4b38f@o4511625394061312.ingest.de.sentry.io/4511625402712144"
    }

    fun initialize() {
        if (!isEnabled()) return
        gateway.start(DSN)
    }

    fun shutdown() {
        gateway.close()
    }

    fun capture(error: Throwable) {
        if (!isEnabled()) return
        gateway.capture(error)
    }

    fun breadcrumb(message: String, category: String = "wallet") {
        if (!isEnabled()) return
        gateway.breadcrumb(message, category)
    }
}

private class AndroidSentryGateway(private val context: Context) : SentryGateway {
    override fun start(dsn: String) {
        SentryAndroid.init(context) { options ->
            options.dsn = dsn
            options.isSendDefaultPii = false
            options.isAttachScreenshot = false
            options.isAttachViewHierarchy = false
            options.isEnableAutoSessionTracking = true
            options.tracesSampleRate = 0.1
            options.profilesSampleRate = 0.0
        }
    }

    override fun close() {
        Sentry.close()
    }

    override fun capture(error: Throwable) {
        Sentry.captureException(error)
    }

    override fun breadcrumb(message: String, category: String) {
        val crumb = Breadcrumb()
        crumb.message = message
        crumb.category = category
        crumb.level = SentryLevel.INFO
        Sentry.addBreadcrumb(crumb)
    }
}

File: ios/CashuWallet/Core/SentryService.swift
import Foundation
import Sentry

enum SentryService {
    // Replace with your DSN from sentry.io → Settings → Projects → apple-ios → Client Keys (DSN)
    private static let dsn =
        "https://aff293071a9e53305e76990761d4b38f@o4511625394061312.ingest.de.sentry.io/4511625402712144"

    static func initialize() {
        guard SettingsStore.shared.sentryEnabled else { return }
        SentrySDK.start { options in
            options.dsn = Self.dsn
            options.sendDefaultPii = false
            #if os(iOS)
            // UIKit-only capture options; sentry-cocoa does not expose them on
            // macOS. Both are off anyway — they are set explicitly so a future
            // SDK default flipping to `true` cannot start shipping screenshots
            // of a wallet.
            options.attachScreenshot = false
            options.attachViewHierarchy = false
            #endif
            options.enableAutoSessionTracking = true
            options.tracesSampleRate = 0.1
        }
    }

    static func shutdown() {
        SentrySDK.close()
    }

    static func capture(_ error: Error) {
        guard SettingsStore.shared.sentryEnabled else { return }
        SentrySDK.capture(error: CrashReportSanitizer.error(error))
    }

    static func breadcrumb(_ message: String, category: String = "wallet") {
        guard SettingsStore.shared.sentryEnabled else { return }
        let crumb = Breadcrumb()
        crumb.message = CrashReportSanitizer.message(message)
        crumb.category = category
        crumb.level = .info
        SentrySDK.addBreadcrumb(crumb)
    }
}

/// Removes known wallet secrets and payment payloads at the explicit crash-report boundary.
///
/// Keep this independent of Sentry so the redaction contract can be unit tested without
/// starting the SDK. Unknown error metadata is intentionally discarded: `NSError.userInfo`
/// may contain the same request data that caused the failure.
enum CrashReportSanitizer {
    private struct Replacement {
        let expression: NSRegularExpression
        let template: String

        init(_ pattern: String, template: String, caseInsensitive: Bool = true) {
            expression = try! NSRegularExpression(
                pattern: pattern,
                options: caseInsensitive ? [.caseInsensitive] : []
            )
            self.template = template
        }
    }

    private static let replacements = [
        Replacement(#"\bnostr\+walletconnect://[^\s,;)\"']+"#, template: "<redacted-nwc-uri>"),
        Replacement(#"\bnsec1[023456789acdefghjklmnpqrstuvwxyz]+\b"#, template: "<redacted-nsec>"),
        Replacement(#"\bcashu[ab][a-z0-9_\-=]{16,}\b"#, template: "<redacted-cashu-token>"),
        Replacement(#"\bcreq(?:a|b1)[a-z0-9_\-=]{8,}\b"#, template: "<redacted-cashu-request>"),
        Replacement(#"\b(?:lnbc|lntb|lnbcrt|lno|lni|lnr|lnurl)[a-z0-9]{16,}\b"#, template: "<redacted-lightning-payload>"),
        Replacement(#"\bbitcoin:(?://)?[^\s,;)\"']+"#, template: "<redacted-bitcoin-uri>"),
        Replacement(#"\b(?:bc1|tb1|bcrt1)[a-z0-9]{20,}\b"#, template: "<redacted-bitcoin-address>"),
        Replacement(#"\b[13mn2][a-km-zA-HJ-NP-Z1-9]{25,34}\b"#, template: "<redacted-bitcoin-address>", caseInsensitive: false),
        Replacement(#"\b[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}\b"#, template: "<redacted-email>"),
        Replacement(#"https?://[^\s,;)\"']+"#, template: "<redacted-url>"),
        Replacement(#"(?<![A-Za-z0-9])/(?:Users|private|data|var|tmp|storage|sdcard)/[^\s,;)\"']+"#, template: "<redacted-path>", caseInsensitive: false),
        Replacement(
            #"\b(mnemonic|seed phrase|private key|secret)\s*[:=]\s*([^\s,;]+(?:\s+[^\s,;]+){0,23})"#,
            template: "$1=<redacted>"
        ),
    ]

    static func message(_ message: String) -> String {
        replacements.reduce(message) { result, replacement in
            let range = NSRange(result.startIndex..<result.endIndex, in: result)
            return replacement.expression.stringByReplacingMatches(
                in: result,
                range: range,
                withTemplate: replacement.template
            )
        }
    }

    static func error(_ error: Error) -> NSError {
        let original = error as NSError
        let typeName = String(reflecting: type(of: error))
        return NSError(
            domain: "CashuWallet.SanitizedError",
            code: original.code,
            userInfo: [NSLocalizedDescriptionKey: "\(typeName): \(message(original.localizedDescription))"]
        )
    }
}
